text/0010-bft-signatures.md

```diff
@@ -0,0 +1,291 @@
+
+### Attestation for block existence
+
+An approach for block denial of service resistance taken by [SmartBFT](https://arxiv.org/abs/2107.06922) is to connect to (at least)`f+1` ordering service nodes, fetch blocks only from one of them and fetch attestations that blocks exist from the remaining `f` (or more).
+If a block existence attestation for a sequence **i** is found valid but the orderer designated to send blocks withholds block **i** for too long, the role assignment of block source and attestation for block existence source is re-shuffled.
+
+An attestation for block existence is simply a regular block (along with its `2f+1` signatures) pruned from its transactions. The signatures can still be verified despite the transactions being pruned from the block thanks
+to the fact that in Fabric the signing algorithm is applied on the ASN1 encoding of the header block which carries the hash of the concatenation of all transactions:
+
+```
+message BlockHeader {
+    uint64 number = 1; // The position in the blockchain
+    bytes previous_hash = 2; // The hash of the previous block header
+    bytes data_hash = 3; // The hash of the BlockData
+}
+```
+
+Let's consider the two possible outcomes based on the orderer that is selected for block retrieval:
+
+1. If orderer that blocks are pulled from is withholding newly formed blocks, then there is at least one honest orderer that will not withhold attestations of block existence. Hence, a re-shuffling of role assignment will occur until an honest orderer will be the block source.
+2. Else, the orderer that blocks are pulled from is not withholding newly formed blocks. However, it is possible that new blocks are not formed due to lack of client activity. It is imperative to ensure that malicious ordering nodes cannot make honest ordering nodes be falsely suspected of block withholding.
+As each "block existence attestation" contains `2f+1` signatures it is impossible to forge one. Hence, if a node sends a attestation for block existence, it means this a block with such sequence number was indeed appended to the blockchain.
+
+In order for a peer or an orderer to convey the intent of receiving only blocks without transactions, 
+the [SeekInfo](https://github.com/hyperledger/fabric-protos/blob/main/orderer/ab.proto#L45) message is expanded to contain also a `SeekContentType` field:
+
+```
+  // SeekContentType indicates what type of content to deliver in response to a request. If BLOCK is specified,
+  // the orderer will stream blocks back to the peer. This is the default behavior. If HEADER_WITH_SIG is  specified, the
+  // orderer will stream only a the header and the signature, and the payload field will be set to nil. This allows
+  // the requester to ascertain that the respective signed block exists in the orderer (or cluster of orderers).
+  enum SeekContentType {
+      BLOCK = 0;
+      HEADER_WITH_SIG =1;
```

Comment:
If you only have block numbers, you need to collect `f+1` attestations before understanding you have enough attestations for a block sequence. Any lower number and you might get fake attestations that will cause you to blame the current block source.

With signatures, you only need to collect one attestation, so the logic is simpler (no need to "count" and track sequences).

Reply:
Yes, you will need to hear from `f+1` for block number (that anyways you would have to connect to either way). Not sure about the logic being simpler v/s difficult as even in the signature way, you would have to assume that f attestations may not be correct. I guess, it would be choice between perform signature matching over attestation (mostly one at run-time but code for handling from all failure) v/s just taking hint from `f+1`